index.js

```diff
@@ -0,0 +1,77 @@
+const http = require("http");
+const net = require("net");
+const dns = require('dns');
+
+const { URL } = require("url");
+
+const proxy = http.createServer(onRequest);
+
+function onRequest(client_req, client_res) {
+  const { port, hostname } = new URL(client_req.url);
+
+  var options = {
+    hostname: hostname,
+    port: port,
+    path: client_req.url,
+    method: client_req.method,
+    headers: client_req.headers,
+  };
+
+  var req = http.request(options, function (res) {
+    client_res.writeHead(res.statusCode, res.headers);
+    res.pipe(client_res, {
+      end: true,
+    });
+  });
+
+  client_req.pipe(req, {
+    end: true,
+  });
+
+  req.on('error', (error) => {
+      console.log('In HTTP Req');
+      console.error(error)
+  })
+}
+
+proxy.on("connect", (req, clientSocket, head) => {
+  // Connect to an origin server
+  const { port, hostname } = new URL(`http://${req.url}`);
+
+  dns.lookup(hostname, (err, address) => {
+      // needed as Pi Hole converts all black listed ips to localhost
+    if (( !err && address !== '0.0.0.0') && hostname !== '0.0.0.0') {
+        const serverSocket = net.connect(port || 80, hostname, () => {
+            clientSocket.write(
+            "HTTP/1.1 200 Connection Established\r\n" +
+                "Proxy-agent: Node.js-Proxy\r\n" +
+                "\r\n"
```

Comment:
Didn't understand the actual use-case of performing dns check here

Reply:
Is it to safeguard proxying to itself?

Reply:
ah no
i have pi hole
that redirects blacklisted domains to 0.0.0.0
which crashes the server as there is no server at 0.0.0.0:80 or :443

Reply:
Got it.